### PR TITLE
Add YouTube client for 1080p muxed videos

### PR DESCRIPTION
**Pull request by clienthax**
### Description of your *pull request* and other information

Adds the ability for yt-dlp to return premuxed 1080p results for videos when available.

image

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check all of the following options that apply:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Fix or improvement to an extractor (Make sure to add/update tests)
- [ ] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

</details>

### CODE REVIEW

**Comment by bashonly** [comment]
I just tested this out, and this started happening to me on the 3rd video:
```
WARNING: [youtube] Skipping player responses from android clients (got player responses for video "aQvGIIdgFDM" instead of "BaW_jenozKc")
```

Definitely should not be in the default clients

**Comment by bashonly** [comment]
Before I got blocked by youtube (?) I was going to write that I think adding another default request to `manifest.googlevideo.com` is probably unwise and will get people rate-limited

**Comment by clienthax** [comment]
I'm not too concerned about it being on by default or not, i just added it there while testing with vrchat to check the returned streams were compatible with avpro easier, as long as it can be enabled with a command line option I'm fine with it not being on defaults.

**Comment by rredford6** [comment]
> I just tested this out, and this started happening to me on the 3rd video:
> 
> ```
> WARNING: [youtube] Skipping player responses from android clients (got player responses for video "aQvGIIdgFDM" instead of "BaW_jenozKc")
> ```
> 
> Note that while it says "android clients", it was for the media_connect_frontend client API request.
> 
> Maybe unrelated, but this has never happened to me before. IMO should not be in the default clients

In the last 12 hours I have seen that on two different videos, but there are other videos that don't produce this warning.  `aQvGIIdgFDM` was claimed to be returned in both instances.  I am using 2024.03.09.155247.  This should probably be tracked as a separate issue.

**Comment by bashonly** [comment]
@rredford6 I only get the bad response with the new client. It pertains to this PR

**Comment by rredford6** [comment]
So then how am I able to reproduce with 2024.03.09.155247?

**Comment by bashonly** [comment]
@rredford6 perhaps you should open a new issue then

**Comment by pukkandan** [comment]
"Wrong player response" issue has been confirmed to not be due to this PR and should be tracked separately

**Comment by coletdjnz** [comment]
You can get the 1080p muxed videos by using a safari user agent with the web client - no need for another client.

I'm guessing media_connect_frontend is android - the android clients are starting to require droidguard, making them unusable for yt-dlp. We will probably be removing them in due course.

**Review by coletdjnz — APPROVED** [review]

File: yt_dlp/extractor/youtube.py [hunk 1]
```diff
@@ -258,7 +268,7 @@ def build_innertube_clients():
     THIRD_PARTY = {
         'embedUrl': 'https://www.youtube.com/',  # Can be any valid URL
     }
-    BASE_CLIENTS = ('ios', 'android', 'web', 'tv', 'mweb')
+    BASE_CLIENTS = ('ios', 'android', 'web', 'tv', 'mweb', 'media_connect_frontend')
```
Comment (coletdjnz) [inline yt_dlp/extractor/youtube.py:271]:
```suggestion
    BASE_CLIENTS = ('ios', 'android', 'web', 'tv', 'mweb')
```

**Review by coletdjnz — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/youtube.py [hunk 2]
```diff
@@ -3619,7 +3629,7 @@ def _extract_player_response(self, client, video_id, master_ytcfg, player_ytcfg,
 
     def _get_requested_clients(self, url, smuggled_data):
         requested_clients = []
-        default = ['ios', 'android', 'web']
+        default = ['ios', 'android', 'web', 'media_connect_frontend']
```
Comment (bashonly) [inline yt_dlp/extractor/youtube.py:3632]:
```suggestion
        default = ['ios', 'android', 'web']
```

**Comment by clienthax** [comment]
> You can also get the 1080p muxed videos by using a safari user agent with the web client, so could also add a duplicate of web client with the safari user agent.
> 
> ~I'm guessing media_connect_frontend is android - the android clients are starting to require droidguard, making them unusable for yt-dlp. We will probably be removing them soon.~ nvm, it seems fine. Worth adding anyways as a backup.

Interesting, is it the same streams as this returns?

**Comment by coletdjnz** [comment]
web client only, using safari UA:

```
[info] Available formats for i1csLh-0L9E:
ID  EXT   RESOLUTION FPS CH │   FILESIZE    TBR PROTO │ VCODEC           VBR ACODEC      ABR ASR MORE INFO
───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
sb3 mhtml 48x27        1    │                   mhtml │ images                                   storyboard
sb2 mhtml 80x45        1    │                   mhtml │ images                                   storyboard
sb1 mhtml 160x90       1    │                   mhtml │ images                                   storyboard
sb0 mhtml 320x180      1    │                   mhtml │ images                                   storyboard
249 webm  audio only      2 │  965.01KiB    47k https │ audio only           opus        47k 48k [en] low, webm_dash
250 webm  audio only      2 │    1.25MiB    62k https │ audio only           opus        62k 48k [en] low, webm_dash
140 m4a   audio only      2 │    2.53MiB   126k https │ audio only           mp4a.40.2  126k 44k [en] medium, m4a_dash
251 webm  audio only      2 │    2.49MiB   125k https │ audio only           opus       125k 48k [en] medium, webm_dash
394 mp4   256x144     24    │    1.38MiB    69k https │ av01.0.00M.08    69k video only          144p, mp4_dash
160 mp4   256x144     24    │    1.96MiB    98k https │ avc1.4d400c      98k video only          144p, mp4_dash
91  mp4   256x144     24    │ ~  3.40MiB   170k m3u8  │ avc1.4D400C          mp4a.40.5           [en]
278 webm  256x144     24    │    1.62MiB    81k https │ vp9              81k video only          144p, webm_dash
395 mp4   426x240     24    │    3.18MiB   159k https │ av01.0.00M.08   159k video only          240p, mp4_dash
133 mp4   426x240     24    │    4.36MiB   219k https │ avc1.4d4015     219k video only          240p, mp4_dash
92  mp4   426x240     24    │ ~  6.22MiB   310k m3u8  │ avc1.4D4015          mp4a.40.5           [en]
242 webm  426x240     24    │    3.66MiB   183k https │ vp9             183k video only          240p, webm_dash
396 mp4   640x360     24    │    6.57MiB   329k https │ av01.0.01M.08   329k video only          360p, mp4_dash
134 mp4   640x360     24    │   10.65MiB   533k https │ avc1.4d401e     533k video only          360p, mp4_dash
93  mp4   640x360     24    │ ~ 16.06MiB   802k m3u8  │ avc1.4D401E          mp4a.40.2           [en]
18  mp4   640x360     24  2 │   14.64MiB   733k https │ avc1.42001E          mp4a.40.2       44k [en] 360p
243 webm  640x360     24    │    7.79MiB   390k https │ vp9             390k video only          360p, webm_dash
397 mp4   854x480     24    │   11.19MiB   561k https │ av01.0.04M.08   561k video only          480p, mp4_dash
135 mp4   854x480     24    │   18.54MiB   929k https │ avc1.4d401e     929k video only          480p, mp4_dash
94  mp4   854x480     24    │ ~ 27.09MiB  1353k m3u8  │ avc1.4D401E          mp4a.40.2           [en]
244 webm  854x480     24    │   13.83MiB   693k https │ vp9             693k video only          480p, webm_dash
398 mp4   1280x720    24    │   20.96MiB  1050k https │ av01.0.05M.08  1050k video only          720p, mp4_dash
136 mp4   1280x720    24    │   32.72MiB  1639k https │ avc1.4d401f    1639k video only          720p, mp4_dash
95  mp4   1280x720    24    │ ~ 51.47MiB  2571k m3u8  │ avc1.4D401F          mp4a.40.2           [en]
247 webm  1280x720    24    │   28.49MiB  1427k https │ vp9            1427k video only          720p, webm_dash
399 mp4   1920x1080   24    │   43.37MiB  2173k https │ av01.0.08M.08  2173k video only          1080p, mp4_dash
137 mp4   1920x1080   24    │   76.01MiB  3807k https │ avc1.640028    3807k video only          1080p, mp4_dash
96  mp4   1920x1080   24    │ ~ 94.57MiB  4724k m3u8  │ avc1.640028          mp4a.40.2           [en]
248 webm  1920x1080   24    │   49.01MiB  2455k https │ vp9            2455k video only          1080p, webm_dash
400 mp4   2560x1440   24    │  109.40MiB  5480k https │ av01.0.12M.08  5480k video only          1440p, mp4_dash
271 webm  2560x1440   24    │  149.59MiB  7493k https │ vp9            7493k video only          1440p, webm_dash
401 mp4   3840x2160   24    │  214.80MiB 10760k https │ av01.0.12M.08 10760k video only          2160p, mp4_dash
313 webm  3840x2160   24    │  304.37MiB 15246k https │ vp9           15246k video only          2160p, webm_dash
```
media connect frontend only:

```
[info] Available formats for i1csLh-0L9E:
ID  EXT   RESOLUTION FPS CH │  FILESIZE   TBR PROTO │ VCODEC        VBR ACODEC      ABR ASR MORE INFO
─────────────────────────────────────────────────────────────────────────────────────────────────────────────────
sb3 mhtml 48x27        1    │                 mhtml │ images                                storyboard
sb2 mhtml 80x45        1    │                 mhtml │ images                                storyboard
sb1 mhtml 160x90       1    │                 mhtml │ images                                storyboard
sb0 mhtml 320x180      1    │                 mhtml │ images                                storyboard
140 m4a   audio only      2 │   2.53MiB  126k https │ audio only        mp4a.40.2  126k 44k [en] medium, m4a_dash
160 mp4   256x144     24    │   1.96MiB   98k https │ avc1.4d400c   98k video only          144p, mp4_dash
91  mp4   256x144     24    │ ~ 3.40MiB  170k m3u8  │ avc1.4d400c       mp4a.40.5           [en]
133 mp4   426x240     24    │   4.36MiB  219k https │ avc1.4d4015  219k video only          240p, mp4_dash
92  mp4   426x240     24    │ ~ 6.22MiB  310k m3u8  │ avc1.4d4015       mp4a.40.5           [en]
134 mp4   640x360     24    │  10.65MiB  533k https │ avc1.4d401e  533k video only          360p, mp4_dash
93  mp4   640x360     24    │ ~16.06MiB  802k m3u8  │ avc1.4d401e       mp4a.40.2           [en]
18  mp4   640x360     24  2 │  14.64MiB  733k https │ avc1.42001E       mp4a.40.2       44k [en] 360p
135 mp4   854x480     24    │  18.54MiB  929k https │ avc1.4d401e  929k video only          480p, mp4_dash
94  mp4   854x480     24    │ ~27.09MiB 1353k m3u8  │ avc1.4d401e       mp4a.40.2           [en]
22  mp4   1280x720    24  2 │ ≈35.33MiB 1765k https │ avc1.64001F       mp4a.40.2       44k [en] 720p
136 mp4   1280x720    24    │  32.72MiB 1639k https │ avc1.4d401f 1639k video only          720p, mp4_dash
95  mp4   1280x720    24    │ ~51.47MiB 2571k m3u8  │ avc1.4d401f       mp4a.40.2           [en]
137 mp4   1920x1080   24    │  76.01MiB 3807k https │ avc1.640028 3807k video only          1080p, mp4_dash
96  mp4   1920x1080   24    │ ~94.57MiB 4724k m3u8  │ avc1.640028       mp4a.40.2           [en]
```

(no need to add safari web client version atm though, can use --user-agent for that)

**Review by coletdjnz — APPROVED** [review]
lgtm

**Comment by clienthax** [comment]
> web client only, using safari UA:
> 
> ```
> [info] Available formats for i1csLh-0L9E:
> ID  EXT   RESOLUTION FPS CH │   FILESIZE    TBR PROTO │ VCODEC           VBR ACODEC      ABR ASR MORE INFO
> ───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
> sb3 mhtml 48x27        1    │                   mhtml │ images                                   storyboard
> sb2 mhtml 80x45        1    │                   mhtml │ images                                   storyboard
> sb1 mhtml 160x90       1    │                   mhtml │ images                                   storyboard
> sb0 mhtml 320x180      1    │                   mhtml │ images                                   storyboard
> 249 webm  audio only      2 │  965.01KiB    47k https │ audio only           opus        47k 48k [en] low, webm_dash
> 250 webm  audio only      2 │    1.25MiB    62k https │ audio only           opus        62k 48k [en] low, webm_dash
> 140 m4a   audio only      2 │    2.53MiB   126k https │ audio only           mp4a.40.2  126k 44k [en] medium, m4a_dash
> 251 webm  audio only      2 │    2.49MiB   125k https │ audio only           opus       125k 48k [en] medium, webm_dash
> 394 mp4   256x144     24    │    1.38MiB    69k https │ av01.0.00M.08    69k video only          144p, mp4_dash
> 160 mp4   256x144     24    │    1.96MiB    98k https │ avc1.4d400c      98k video only          144p, mp4_dash
> 91  mp4   256x144     24    │ ~  3.40MiB   170k m3u8  │ avc1.4D400C          mp4a.40.5           [en]
> 278 webm  256x144     24    │    1.62MiB    81k https │ vp9              81k video only          144p, webm_dash
> 395 mp4   426x240     24    │    3.18MiB   159k https │ av01.0.00M.08   159k video only          240p, mp4_dash
> 133 mp4   426x240     24    │    4.36MiB   219k https │ avc1.4d4015     219k video only          240p, mp4_dash
> 92  mp4   426x240     24    │ ~  6.22MiB   310k m3u8  │ avc1.4D4015          mp4a.40.5           [en]
> 242 webm  426x240     24    │    3.66MiB   183k https │ vp9             183k video only          240p, webm_dash
> 396 mp4   640x360     24    │    6.57MiB   329k https │ av01.0.01M.08   329k video only          360p, mp4_dash
> 134 mp4   640x360     24    │   10.65MiB   533k https │ avc1.4d401e     533k video only          360p, mp4_dash
> 93  mp4   640x360     24    │ ~ 16.06MiB   802k m3u8  │ avc1.4D401E          mp4a.40.2           [en]
> 18  mp4   640x360     24  2 │   14.64MiB   733k https │ avc1.42001E          mp4a.40.2       44k [en] 360p
> 243 webm  640x360     24    │    7.79MiB   390k https │ vp9             390k video only          360p, webm_dash
> 397 mp4   854x480     24    │   11.19MiB   561k https │ av01.0.04M.08   561k video only          480p, mp4_dash
> 135 mp4   854x480     24    │   18.54MiB   929k https │ avc1.4d401e     929k video only          480p, mp4_dash
> 94  mp4   854x480     24    │ ~ 27.09MiB  1353k m3u8  │ avc1.4D401E          mp4a.40.2           [en]
> 244 webm  854x480     24    │   13.83MiB   693k https │ vp9             693k video only          480p, webm_dash
> 398 mp4   1280x720    24    │   20.96MiB  1050k https │ av01.0.05M.08  1050k video only          720p, mp4_dash
> 136 mp4   1280x720    24    │   32.72MiB  1639k https │ avc1.4d401f    1639k video only          720p, mp4_dash
> 95  mp4   1280x720    24    │ ~ 51.47MiB  2571k m3u8  │ avc1.4D401F          mp4a.40.2           [en]
> 247 webm  1280x720    24    │   28.49MiB  1427k https │ vp9            1427k video only          720p, webm_dash
> 399 mp4   1920x1080   24    │   43.37MiB  2173k https │ av01.0.08M.08  2173k video only          1080p, mp4_dash
> 137 mp4   1920x1080   24    │   76.01MiB  3807k https │ avc1.640028    3807k video only          1080p, mp4_dash
> 96  mp4   1920x1080   24    │ ~ 94.57MiB  4724k m3u8  │ avc1.640028          mp4a.40.2           [en]
> 248 webm  1920x1080   24    │   49.01MiB  2455k https │ vp9            2455k video only          1080p, webm_dash
> 400 mp4   2560x1440   24    │  109.40MiB  5480k https │ av01.0.12M.08  5480k video only          1440p, mp4_dash
> 271 webm  2560x1440   24    │  149.59MiB  7493k https │ vp9            7493k video only          1440p, webm_dash
> 401 mp4   3840x2160   24    │  214.80MiB 10760k https │ av01.0.12M.08 10760k video only          2160p, mp4_dash
> 313 webm  3840x2160   24    │  304.37MiB 15246k https │ vp9           15246k video only          2160p, webm_dash
> ```
> 
> media connect frontend only:
> 
> ```
> [info] Available formats for i1csLh-0L9E:
> ID  EXT   RESOLUTION FPS CH │  FILESIZE   TBR PROTO │ VCODEC        VBR ACODEC      ABR ASR MORE INFO
> ─────────────────────────────────────────────────────────────────────────────────────────────────────────────────
> sb3 mhtml 48x27        1    │                 mhtml │ images                                storyboard
> sb2 mhtml 80x45        1    │                 mhtml │ images                                storyboard
> sb1 mhtml 160x90       1    │                 mhtml │ images                                storyboard
> sb0 mhtml 320x180      1    │                 mhtml │ images                                storyboard
> 140 m4a   audio only      2 │   2.53MiB  126k https │ audio only        mp4a.40.2  126k 44k [en] medium, m4a_dash
> 160 mp4   256x144     24    │   1.96MiB   98k https │ avc1.4d400c   98k video only          144p, mp4_dash
> 91  mp4   256x144     24    │ ~ 3.40MiB  170k m3u8  │ avc1.4d400c       mp4a.40.5           [en]
> 133 mp4   426x240     24    │   4.36MiB  219k https │ avc1.4d4015  219k video only          240p, mp4_dash
> 92  mp4   426x240     24    │ ~ 6.22MiB  310k m3u8  │ avc1.4d4015       mp4a.40.5           [en]
> 134 mp4   640x360     24    │  10.65MiB  533k https │ avc1.4d401e  533k video only          360p, mp4_dash
> 93  mp4   640x360     24    │ ~16.06MiB  802k m3u8  │ avc1.4d401e       mp4a.40.2           [en]
> 18  mp4   640x360     24  2 │  14.64MiB  733k https │ avc1.42001E       mp4a.40.2       44k [en] 360p
> 135 mp4   854x480     24    │  18.54MiB  929k https │ avc1.4d401e  929k video only          480p, mp4_dash
> 94  mp4   854x480     24    │ ~27.09MiB 1353k m3u8  │ avc1.4d401e       mp4a.40.2           [en]
> 22  mp4   1280x720    24  2 │ ≈35.33MiB 1765k https │ avc1.64001F       mp4a.40.2       44k [en] 720p
> 136 mp4   1280x720    24    │  32.72MiB 1639k https │ avc1.4d401f 1639k video only          720p, mp4_dash
> 95  mp4   1280x720    24    │ ~51.47MiB 2571k m3u8  │ avc1.4d401f       mp4a.40.2           [en]
> 137 mp4   1920x1080   24    │  76.01MiB 3807k https │ avc1.640028 3807k video only          1080p, mp4_dash
> 96  mp4   1920x1080   24    │ ~94.57MiB 4724k m3u8  │ avc1.640028       mp4a.40.2           [en]
> ```
> 
> (no need to add safari web client version atm though, can use --user-agent for that)

Hm.. I can't seem to reproduce this, Am I doing something wrong here?
`yt-dlp  https://www.youtube.com/watch?v=i1csLh-0L9E --list-formats --user-agent "Mozilla/5.0 (Macintosh; Intel Mac OS X 14_4_1) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/17.3.1 Safari/605.1.15" --extractor-args "youtube:player_client=web"`
image

**Comment by bashonly** [comment]
try `--impersonate safari`

**Comment by clienthax** [comment]
> try `--impersonate safari` instead of passing UA

Yep, works with that, thanks

**Comment by clienthax** [comment]
Is there a way to only impersonate safari for a specific extractor?

**Comment by bashonly** [comment]
there will be when #9474 is merged

**Review by bashonly — APPROVED** [review]

File: yt_dlp/extractor/youtube.py [hunk 3]
```diff
@@ -238,6 +238,16 @@
         },
         'INNERTUBE_CONTEXT_CLIENT_NAME': 85
     },
+    # This client has 1080p non split stream videos
+    'media_connect_frontend': {
```
Comment (bashonly) [inline yt_dlp/extractor/youtube.py:242]: as pukkandan mentioned on discord, `_` has special meaning in client names, so we need to change this to something `mediaconnect`

**Review by bashonly — CHANGES_REQUESTED** [review]

**Review by bashonly — APPROVED** [review]

**Comment by Elite** [comment]
This ``yt-dlp -gf 96 tIIWYh601tg --impersonate safari`` gives me .m3u8 playlist and not the direct mp4 link. What am I missing here?

**Comment by bashonly** [comment]
@Elite safari is served m3u8 formats. don't pass `--impersonate safari`

**Comment by Elite** [comment]
Removing that also results in same!

[2024-06-07_09-06](https://github.com/yt-dlp/yt-dlp/assets/5461439/760cb737-3861-485b-b060-89f857834a7c)

**Comment by bashonly** [comment]
@Elite the `mediaconnect` client is also served m3u8 formats. Why are you using it if you want http mp4 formats? and you are passing `-f 96` which will always select an m3u8 format

**Comment by seproDev** [comment]
@Elite I think you are conflating being pre-merged and the protocol used. Format 96 is pre-merged (audio and video are together), but it is transmitted with segments in a m3u8 and not as a single mp4 file. Afaik. there is no pre-merged http mp4 URL.

**Comment by Elite** [comment]
@seproDev Yes, that was the expectation getting premuxed mp4 over http, thanks for clarifying.

**Comment by Elite** [comment]
One question though, is this format 96 hls stream downloadable only from the IP it was generated or can it be used from other IP as well?

**Comment by bashonly** [comment]
@Elite depends on the video. The original requestor's IP address is always embedded in the URL, but the regional/IP restriction is not always enforced

**Pull request merged**